### ISSUE TITLE
Http cache not caching always

### ISSUE THREAD

**Issue by anup1245** [apollographql/apollo-android#804]
In http cache its seen that, the query gets cached first time, and when the same query is executed again, the returned data is from network and not cache.
But sometimes its returned from cache and if consecutively calls are made then sometimes cache throws an error and network data is returned.

If data is cached, when is it removed from cache, I'm not setting any expiration date for it
Using CACHE_FIRST flag

**Comment by sav007** [comment]
>cache throws an error and network data is returned

Could you please catch the stack trace of the exception?

What version of apollo client are you using?

**Comment by anup1245** [comment]
@sav007 I was trying to achieve CACHE_AND_NETWORK using Http cache by concatenating 2 observables, this maybe  causing the issue.

Is there a way to achieve CACHE_AND_NETWORK using http cache ?

**Comment by sav007** [comment]
We don't support right now `CACHE_AND_NETWORK ` for http cache, but I think it's a good feature request. First try to use the latest version, second you can try to use normalized cache (but without http cache) that has such policy.

**Comment by fluxxion82** [comment]
I'm having all sorts of issues with the caching. Like it was mentioned above, the query results sometimes get pulled from the cache despite the policy being Network first and there is no difference even when I explicitly clear the caches with `clearNormalizedCache()` and `clearHttpCache()`. I've tried using one cache at a time and even no cache at all, and the results are totally inconsistent and seemingly random. Things will usually work the first time I make the call, but subsequent calls are incorrect. I will consistently get 404 errors after the first successful attempt. I'm using v0.4.4

**Comment by fluxxion82** [comment]
I think my issues was having the network interceptors being added to the OkHttpClient with 'addInterceptor' instead of 'addNetworkInterceptor'. Things seem better now that I made that change.

**Comment by anup1245** [comment]
@fluxxion82 `addNetworkInterceptor` fixed the issue ?

**Comment by fluxxion82** [comment]
@anup1245 No. Well, I moved one of my interceptors to the `addNetworkInterceptor` and I believe it's working fine there. The other interceptor I had to keep using with `addInterceptor`. I had a lot of issues at first but I later figured out how to get everything working. It could have been that I didn't fully understand how to use the interceptor, but for every network call I make (rest or graphql), I have to configure the interceptor to make sure it has all of the correct headers and url etc.
